Work around __make_dynamic_extent compiler bug

mhoemmen opened the pull request:
Some compilers can't do CTAD correctly with extents.  Replacing `__make_dynamic_extent<IndexTypes>()` with an explicit expression in the deduction guide's parameter pack expansion works around the issue.

The resulting change means that nothing in mdspan uses `__make_dynamic_extent`.  I've left `__make_dynamic_extent` defined, just in case downstream projects rely on it.

Fixes #177.  @youyu3

mhoemmen committed:
Work around __make_dynamic_extent compiler bug

Some compilers can't do CTAD correctly with extents.  Replacing
__make_dynamic_extent<IndexTypes>() with an explicit expression in the
deduction guide's parameter pack expansion works around the issue.

The resulting change means that nothing in mdspan uses
__make_dynamic_extent.  I've left __make_dynamic_extent defined,
just in case downstream projects rely on it.

Fixes #177.

crtrott commented:
Lets delete that thing. Nobody should start to rely on things in detail with two underscores as start characters. Better they learn that now.

mhoemmen committed:
Remove __make_dynamic_extent(_integral)

Remove __make_dynamic_extent and __make_dynamic_extent_integral,
as they are no longer needed.

This relates to #177.

mhoemmen commented:
@crtrott I removed `__make_dynamic_extent` and `__make_dynamic_extent_integral` -- thanks!
